store_init.go

```diff
@@ -0,0 +1,278 @@
+	err = dbPool.QueryRow(ctx, `SELECT CURRENT_SETTING('server_version_num')::INTEGER`).Scan(&currentPgVersion)
+	if err != nil {
+		return nil, xerrors.Errorf("retrieving server version failed: %w", err)
+	}
+	if currentPgVersion < minimumPgVersion {
+		return nil, fmt.Errorf(
+			"code was tested on PostgreSQL version %d only: change the source if you want to try running on the older version %d",
+			minimumPgVersion, currentPgVersion,
+		)
+	}
+
+	dbSettingsExpected := "Encoding:SQL_ASCII Collate:C Ctype:C"
+	var dbSettings string
+	if err := dbPool.QueryRow(
+		ctx,
+		`
+		SELECT 'Encoding:' || pg_encoding_to_char(encoding) || ' Collate:' || datcollate || ' Ctype:' || datctype
+			FROM pg_database
+		WHERE datname = current_database()
+		`,
+	).Scan(&dbSettings); err != nil {
+		return nil, err
+	}
+	if dbSettings != dbSettingsExpected {
+		return nil, fmt.Errorf(
+			"unexpected database settings: you must create your database with something like `%s` for reliable and performant binary storage\n Current settings: %s\nExpected settings: %s",
+			`CREATE DATABASE {{name}} ENCODING='SQL_ASCII' LC_COLLATE='C' LC_CTYPE='C' TEMPLATE='template0'`,
+			dbSettings, dbSettingsExpected,
+		)
+	}
+
+	// check whether we are talking to a slave, and force RO to shut off tipset-tracking automatically
+	if dbbs.isWritable {
+		if _, err := dbPool.Exec(ctx, fmt.Sprintf(
+			"CREATE TEMPORARY TABLE %s ( pk INTEGER ) ON COMMIT DROP",
+			"tmptable_"+randBytesAsHex(),
+		)); err != nil {
+			dbbs.isWritable = false
+		}
+	}
+
+	dbbs.lru, err = cidkeyedlru.NewCidKeyedLruCache(dbbs.lruSizeBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to instantiate cache: %w", err)
+	}
+
+	if cfg.LogDetailedAccess {
+		if dbbs.instanceNamespace == "" {
+			return nil, xerrors.New("detailed access logging not possible without an instance-specific chainsync namespace")
+		}
+		if !dbbs.isWritable {
+			return nil, xerrors.New("detailed access logging not possible on a read-only store")
+		}
+		dbbs.accessLogsDetailed = make(map[accessUnit]int64, 65535)
+	}
+
+	neededDDL, err := dbbs.neededDeployDDL(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to determine current deployed schema status: %w", err)
+	}
+
+	if neededDDL != nil {
+
+		if !dbbs.isWritable {
+			if dbbs.instanceNamespace != "" {
+				return nil, fmt.Errorf(
+					"unable to continue: connecting in ReadOnly mode requires that the view %s.current_head and all associates dependencies have been already deployed",
+					dbbs.instanceNamespace,
+				)
+			}
+			return nil, xerrors.New(
+				"unable to continue: connecting in ReadOnly mode requires that the fil_common_base schema has been already deployed",
+			)
+		}
+
+		if !cfg.AutoUpdateSchema {
+			return nil, xerrors.New(
+				"unable to continue: currently-deployed schema does not (partially) match the expected state, and AutoUpdateSchema is not enabled",
+			)
+		}
+
+		if err = dbbs.deploy(ctx, neededDDL); err != nil {
```

Comment:
This deploy happens before you have gained exclusive access to the database?

Reply:
> exclusive access to the database?

There is no exclusive lock on the db, only on namespaces ( there can and often are multiple writers on the same RDBMS, each with its own namespace )

The entire deploy is guarded by a serializable transaction here: https://github.com/filecoin-project/go-bs-postgres-chainnotated/blob/8b1f2e9296c4346ca7647889cde5d2ecae123f8e/store_schema.go#L569-L592
It will either happen or not at all

Let me know if this doesn't answer the concern in full.